<a href="https://colab.research.google.com/github/sosarodrigox/grupo13_pp2/blob/main/ProyectoIntegrador_Grupo13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Instituto Superior Politécnico Córdoba

## Tecnicatura Superior en Ciencia de Datos e Inteligencia Artificial

### Cátedra: Práctica Profesionalizante II

**Docente**: Facundo Oliva Cuneo

**Año**: 2024
  
**Trabajo Práctico Integrador**

## Grupo 13
### Integrantes:

- **Ingaramo, Ma. Eugenia**  
  [eugenia.ingaramo@gmail.com](mailto:eugenia.ingaramo@gmail.com)
- **Lonardi, Pablo**  
  [lonardipablo@gmail.com](mailto:lonardipablo@gmail.com)
- **Margheim, Carolina**  
  [caro08.m@gmail.com](mailto:caro08.m@gmail.com)
- **Sosa, Rodrigo**  
  [sosarodrigox@gmail.com](mailto:sosarodrigox@gmail.com)
- **Zenere, Mauricio**  
  [zeneremauricio@gmail.com](mailto:zeneremauricio@gmail.com)

### CRISP-DM 1. Compresión del negocio:

### CRISP-DM 2. Compresión de los datos:

In [1]:
import pandas as pd

# URL del archivo en formato 'Raw' en GitHub
url = "https://raw.githubusercontent.com/sosarodrigox/grupo13_pp2/main/data/data.csv"

# Cargar los datos en un DataFrame
data = pd.read_csv(url)

# Mostrar las primeras filas para confirmar la carga correcta
print(data.head())

          isbn13      isbn10           title subtitle  \
0  9780002005883  0002005883          Gilead      NaN   
1  9780002261982  0002261987    Spider's Web  A Novel   
2  9780006163831  0006163831    The One Tree      NaN   
3  9780006178736  0006178731  Rage of angels      NaN   
4  9780006280897  0006280897  The Four Loves      NaN   

                           authors                     categories  \
0               Marilynne Robinson                        Fiction   
1  Charles Osborne;Agatha Christie  Detective and mystery stories   
2             Stephen R. Donaldson               American fiction   
3                   Sidney Sheldon                        Fiction   
4              Clive Staples Lewis                 Christian life   

                                           thumbnail  \
0  http://books.google.com/books/content?id=KQZCP...   
1  http://books.google.com/books/content?id=gA5GP...   
2  http://books.google.com/books/content?id=OmQaw...   
3  http://books.go

In [2]:
# Configurar opciones de visualización
pd.set_option("display.max_columns", None)  # Todas las columnas
pd.set_option("display.max_rows", 10)  # 0 filas

# DataFrame
data

isbn13      isbn10                      title  \
0     9780002005883  0002005883                     Gilead   
1     9780002261982  0002261987               Spider's Web   
2     9780006163831  0006163831               The One Tree   
3     9780006178736  0006178731             Rage of angels   
4     9780006280897  0006280897             The Four Loves   
...             ...         ...                        ...   
6805  9788185300535  8185300534                  I Am that   
6806  9788185944609  8185944601       Secrets Of The Heart   
6807  9788445074879  8445074873             Fahrenheit 451   
6808  9789027712059  9027712050   The Berlin Phenomenology   
6809  9789042003408  9042003405  'I'm Telling You Stories'   

                                            subtitle  \
0                                                NaN   
1                                            A Novel   
2                                                NaN   
3                                                NaN   
4                                                NaN   
...                                              ...   
6805             Talks with Sri Nisargadatta Maharaj   
6806                                             NaN   
6807                                             NaN   
6808                                             NaN   
6809  Jeanette Winterson and the Politics of Reading   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                             Stephen R. Donaldson   
3                                   Sidney Sheldon   
4                              Clive Staples Lewis   
...                                            ...   
6805  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
6806                                 Khalil Gibran   
6807                                  Ray Bradbury   
6808                 Georg Wilhelm Friedrich Hegel   
6809                        Helena Grice;Tim Woods   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                  American fiction   
3                           Fiction   
4                    Christian life   
...                             ...   
6805                     Philosophy   
6806                      Mysticism   
6807                   Book burning   
6808                        History   
6809             Literary Criticism   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=OmQaw...   
3     http://books.google.com/books/content?id=FKo2T...   
4     http://books.google.com/books/content?id=XhQ5X...   
...                                                 ...   
6805  http://books.google.com/books/content?id=Fv_JP...   
6806  http://books.google.com/books/content?id=XcrVp...   
6807                                                NaN   
6808  http://books.google.com/books/content?id=Vy7Sk...   
6809  http://books.google.com/books/content?id=2lVyR...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     Volume Two of Stephen Donaldson's acclaimed se...          1982.0   
3     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
4     Lewis' work on the nature of love divides love...          2002.0   
...                                                 ...             ...   
6805  This collection of the timeless teachings of o...          1999.0   
6806                                                NaN          1993.0   
6807                                                NaN          2004.0   
6808  Since the th

### Analisis de datos EDA:

In [3]:
# Obtener informacion del dataset recibido:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   isbn13          6810 non-null   int64  
 1   isbn10          6810 non-null   object 
 2   title           6810 non-null   object 
 3   subtitle        2381 non-null   object 
 4   authors         6738 non-null   object 
 5   categories      6711 non-null   object 
 6   thumbnail       6481 non-null   object 
 7   description     6548 non-null   object 
 8   published_year  6804 non-null   float64
 9   average_rating  6767 non-null   float64
 10  num_pages       6767 non-null   float64
 11  ratings_count   6767 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 638.6+ KB


### Resumen del Dataset de Libros

El dataset analizado consta de **6,810 entradas**, cada una representa un libro único. Estos están descritos mediante **12 atributos**, que incluyen tanto identificadores numéricos como descriptivos. A continuación se detalla la estructura y las características principales del dataset:

### Identificadores de Libro
- `isbn13`: Cada libro tiene un identificador ISBN de 13 dígitos, proporcionado para todos los libros (6810 entradas no nulas).
- `isbn10`: Un identificador ISBN alternativo de 10 dígitos para cada libro, también completamente poblado.

### Títulos y Autores
- `title`: El título del libro, sin entradas nulas en el dataset.
- `subtitle`: Subtítulos proporcionados para 2,381 libros, sugiriendo que no todos los libros tienen un subtítulo explícito.
- `authors`: Nombres de los autores, presentes para 6,738 libros, indicando una pequeña proporción de entradas sin esta información. Esto podría representar un problema a futuro.

### Categorización y Descripción
- `categories`: Categorías o géneros bajo los cuales se clasifican los libros, con 6,711 libros categorizados. Hay casi 100 libros no categorizados lo cual podria generar conflictos a la hora de realizar recomendaciones.
- `description`: Descripciones textuales detalladas disponibles para 6,548 libros, proporcionando un contexto valioso sobre el contenido y el tema de cada libro y que podria utilizarce para las recomendaciones a clientes de la biblioteca.

### Aspectos Visuales y Editoriales
- `thumbnail`: URLs de imágenes de portada para 6,481 libros, lo que ayuda en la visualización y el marketing digital del libro. Se testeó valores al azar y se verificó que algunos links estaban caidos. Evaluar la solución para estos casos.
- `published_year`: Año de publicación registrado para 6,804 libros, ofreciendo una perspectiva sobre la temporalidad de las publicaciones.

### Calificaciones y Popularidad
- `average_rating`: La calificación promedio asignada por los usuarios, disponible para 6,767 libros, lo que refleja la recepción del libro por parte de los lectores.
- `num_pages`: Número de páginas, también disponible para 6,767 libros, indicando la extensión de los textos.
- `ratings_count`: El número de calificaciones que ha recibido cada libro, también notado para 6,767 libros, que puede ser utilizado para medir la popularidad o el engagement de los lectores.

**Uso del Dataset en Relación con los Objetivos del Proyecto**: Este conjunto de datos ofrece una base sólida para varios componentes esenciales del proyecto. *Es ideal para desarrollar un sistema automatizado de recomendación de libros*, tanto basado en contenido como en colaboración, alineándose con el objetivo de optimizar la experiencia del usuario y aumentar su satisfacción. Sin embargo, para respaldar completamente la iniciativa de mejorar la eficiencia en la gestión de recursos y optimizar la utilización de estos, sería beneficioso enriquecer el dataset con datos adicionales sobre las interacciones de los usuarios con la biblioteca, tales como:

 * Registros de préstamos.
 * Frecuencia de uso.
 * Retroalimentación directa de los usuarios.

Esta información adicional permitiría realizar análisis predictivos más precisos sobre tendencias de publicación y demanda de libros, facilitando una mejor alineación con las necesidades e intereses de los usuarios. Además, la integración y automatización de los registros de disponibilidad y estado de préstamo en el dataset podrían reducir significativamente la carga administrativa del personal, apoyando la eficiencia operativa y proporcionando una base de datos más completa sobre la cual tomar decisiones informadas para la gestión de colecciones y la programación de actividades de la biblioteca.

De todas formas, como se mencionó anteriormente el dataset es **suficiente** y presenta datos de **calidad** para una primera iteración en la construcción de un sistema de recomendación.

### CRISP-DM 3. Preparación de los datos:

Dado que uno de los objetivos es implementar un sistema automatizado de recomendación de libros, vemos que en el dataset contamos con los datos necesarios para identificar un libro y recomendarlo a los usuarios de la biblioteca, procedemos a extraer estas variables importantes y procesarlas.

Las variables seleccionadas son:

* `title` (ítulo del libro)
* `authors` (Autor/es del libro)
* `categories` (Categorías del libro)
* `description` (Descripción del libro)
* `published_year` (Año de publicación)

In [4]:
# Obtenemos las variables de interés para el análisis:

variables_seleccionadas = [
    "title",
    "authors",
    "categories",
    "description",
    "published_year",
]

print(data[variables_seleccionadas].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           6810 non-null   object 
 1   authors         6738 non-null   object 
 2   categories      6711 non-null   object 
 3   description     6548 non-null   object 
 4   published_year  6804 non-null   float64
dtypes: float64(1), object(4)
memory usage: 266.1+ KB
None


Antes de construir el sistema de recomendación, debemos preprocesar los datos. Esto incluye limpieza de texto, manejo de valores faltantes y tokenización.

In [5]:
# Revisar si hay valores nulos en las variables seleccionadas y reemplazarlos por un valor por defecto
for feature in variables_seleccionadas:
    data[feature] = data[feature].fillna("")

# Revisar si hay valores nulos en las variables seleccionadas
print(data[variables_seleccionadas].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   title           6810 non-null   object
 1   authors         6810 non-null   object
 2   categories      6810 non-null   object
 3   description     6810 non-null   object
 4   published_year  6810 non-null   object
dtypes: object(5)
memory usage: 266.1+ KB
None


Ahora que no tenemos valores nulos, combinamos las columnas de interés en una nueva columna llamada `libros`. El sistema de recomendación utilizará esta columna combinada para encontrar similitudes entre los libros de la biblioteca.

In [6]:
# Combinar las columnas de interés en una sola columna
libros = (
    data["title"]
    + " "
    + data["categories"]
    + " "
    + data["authors"]
    + " "
    + f"{data['published_year']}"
)
pd.set_option("display.max_colwidth", None)
libros.head(5)

0                                             Gilead Fiction Marilynne Robinson 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
1    Spider's Web Detective and mystery stories Charles Osborne;Agatha Christie 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
2                            The One Tree American fiction Stephen R. Donaldson 0       2004.0\n1       2000.0\n2       1982.0\n3       1993.0\n4       2002.0\n         ...  \n6805    1999.0\n6806    1993.0\n6807    2004.0\n6808    1981.0\n6809    1998.0\nName: published_year, Length: 6810, dtype: object
3                                         Rage of angels Fiction Sidney Sheldon 0 

### CRISP-DM 4. Modelado:

Para crear el sistema de recomendación basado en contenido, utilizaremos la técnica de **TF-IDF** (Term Frequency-Inverse Document Frequency). Esta técnica se utiliza para convertir texto en vectores numéricos, que luego se pueden utilizar para calcular la similitud entre los textos. TF-IDF otorga más peso a los términos que son importantes en un documento específico y menos peso a los términos comunes.

* Frecuencia de Término (TF): Mide cuántas veces aparece un término en un documento.
* Frecuencia Inversa de Documento (IDF): Mide la importancia de un término en el conjunto de documentos. Los términos comunes que aparecen en muchos documentos reciben un peso menor.

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Crear un vectorizador TF-IDF
vector_tfidf = TfidfVectorizer()

vector_caracteristicas = vector_tfidf.fit_transform(libros)

print(vector_caracteristicas.shape)
print(vector_caracteristicas)

(6810, 10548)
  (0, 6894)	0.06443750710614472
  (0, 2840)	0.06443750710614472
  (0, 143)	0.06443750710614472
  (0, 5557)	0.06443750710614472
  (0, 7629)	0.06443750710614472
  (0, 6655)	0.06443750710614472
  (0, 102)	0.06443750710614472
  (0, 142)	0.06443750710614472
  (0, 91)	0.06443750710614472
  (0, 141)	0.06443750710614472
  (0, 140)	0.06443750710614472
  (0, 139)	0.06443750710614472
  (0, 103)	0.06443750710614472
  (0, 138)	0.06443750710614472
  (0, 110)	0.06443750710614472
  (0, 98)	0.12887501421228945
  (0, 92)	0.06443750710614472
  (0, 108)	0.06443750710614472
  (0, 112)	0.12887501421228945
  (0, 8035)	0.43700046813706117
  (0, 6013)	0.5885172279797457
  (0, 3422)	0.11189605685054443
  (0, 3882)	0.5885172279797457
  (1, 1822)	0.3013956698681646
  (1, 267)	0.30498953565603815
  :	:
  (6809, 10330)	0.33175135989257093
  (6809, 2240)	0.2206043839379868
  (6809, 9492)	0.30218650531542857
  (6809, 10435)	0.26439634843355025
  (6809, 5684)	0.20438577371569946
  (6809, 9065)	0.21319651

**Similitud del Coseno**(Cosine Similarity): La similaridad de coseno es una métrica que se utiliza para medir cuán similares son dos vectores. Se calcula como el coseno del ángulo entre dos vectores en un espacio multidimensional. En términos de recomendación de contenido, esto nos ayuda a medir cuán similares son dos libros en función de sus características textuales vectorizadas.Se calcula tomando el coseno del ángulo entre dos vectores no nulos y su valor oscila entre -1 y 1, donde:

* Un valor de 1 indica que los vectores son idénticos.
* Un valor de 0 indica que los vectores son ortogonales (no tienen similitudes).
* Un valor de -1 indica que los vectores son diametralmente opuestos.
* La diagonal principal de la matriz siempre estará compuesta por unos (1,1,1,...,1), ya que cada libro es idéntico a sí mismo.

En la función cosine_similarity() pasamos el `vector_caracteristicas` dos veces indicando que estamos calculando la similaridad de cada libro con todos los demás libros (incluido él mismo) para generar una matriz de similaridad. En esta matriz, la entrada [i, j] representa la similaridad entre el libro i y el libro j

Ejemplo Visual: Si tenemos tres libros (A, B, C), la matriz de similaridad resultante tendría este aspecto:

        A    B    C
    A  1.0  0.8  0.3
    B  0.8  1.0  0.5
    C  0.3  0.5  1.0

Aquí, cada celda [i, j] representa la similaridad entre el libro i y el libro j, donde el valor de 1.0 en la diagonal indica que un libro es completamente similar a sí mismo.


In [8]:
from sklearn.metrics.pairwise import cosine_similarity

# Obtener la similitud coseno entre los vectores de características:
similitud = cosine_similarity(vector_caracteristicas, vector_caracteristicas)

print(similitud)

[[1.         0.08005311 0.11641123 ... 0.08492469 0.08320896 0.07576596]
 [0.08005311 1.         0.08011208 ... 0.06549297 0.06416982 0.10645452]
 [0.11641123 0.08011208 1.         ... 0.08498725 0.09570742 0.07582177]
 ...
 [0.08492469 0.06549297 0.08498725 ... 1.         0.06807484 0.06198557]
 [0.08320896 0.06416982 0.09570742 ... 0.06807484 1.         0.06073329]
 [0.07576596 0.10645452 0.07582177 ... 0.06198557 0.06073329 1.        ]]


### CRISP-DM 5. Evaluación:

En esta fase, evaluaremos la efectividad del sistema de recomendación. Este paso implica probar el sistema con ejemplos concretos y verificar si las recomendaciones son relevantes.

Pasos a seguir:

* Crear una lista con todos los títulos de libros en el dataset.
* Obtener el nombre del libro del usuario.
* Encontrar la coincidencia más cercana para el nombre del libro dado por el usuario.
* Encontrar el índice del libro con el título coincidente.
* Obtener una lista de libros similares basados en el puntaje de similitud.
* Ordenar los libros basados en su puntaje de similitud.
* Imprimir los nombres de libros similares basados en el índice.

In [9]:
# Crear una lista con todos los títulos de libros en el dataset
titulos = data["title"].tolist()
print(titulos)

['Gilead', "Spider's Web", 'The One Tree', 'Rage of angels', 'The Four Loves', 'The Problem of Pain', 'An Autobiography', 'Empires of the Monsoon', 'The Gap Into Madness', 'Master of the Game', 'If Tomorrow Comes', "Assassin's Apprentice", 'Warhost of Vastmark', 'The Once and Future King', 'Murder in LaMut', 'Jimmy the Hand', 'Well of Darkness', 'Witness for the Prosecution & Selected Plays', 'The Little House', 'Mystical Paths', 'Glittering Images', 'Glamorous Powers', 'The Mad Ship', 'Post Captain', 'The Reverse of the Medal', 'Miss Marple', 'The Years of Rice and Salt', 'Spares', 'Gravity', 'The Wise Woman', "Girls' Night in", 'The White Album', "The Bonesetter's Daughter", 'The Lexus and the Olive Tree', "'Tis", 'Ocean Star Express', 'A Small Pinch of Weather', 'The Princess of the Chalet School', 'Koko', 'Tree and Leaf', 'Partners in Crime', 'Murder in Mesopotamia', 'The Lord of the Rings, the Return of the King', 'All Families are Psychotic', 'Death in the Clouds', 'Appointment w

In [10]:
# Obtener el nombre del libro del usuario.
def obtener_libro_usuario():
    libro_usuario = input("Ingrese el título de un libro: ")
    return libro_usuario

Para encontrar las coincidencias usamos la librerías difflib, esta proporciona clases y funciones para comparar secuencias, y es particularmente útil para tareas de procesamiento de texto y análisis de diferencias. En este caso, usamos la función get_close_matches() para encontrar las coincidencias más cercanas al título del libro dado por el usuario.

In [11]:
from difflib import get_close_matches

# Primero obtenemos el nombre del libro del usuario:
libro_usuario = obtener_libro_usuario()


# Función para encontrar la coincidencia más cercana al libro ingresado por el usuario.
def encontrar_coincidencia(titulo_ingresado):
    return get_close_matches(titulo_ingresado, titulos)


coincidencia = encontrar_coincidencia(libro_usuario)

print(coincidencia)

['Reaper Man', 'Teacher Man']


In [12]:
# Encontrar el índice del libro con el título coincidente.
coincidencia_index = coincidencia[0]
indice_libro = data[data["title"] == coincidencia_index].index[0]

In [13]:
# Obtener una lista de libros similares basados en el puntaje de similitud.
puntaje_similitud = list(enumerate(similitud[indice_libro]))
# print(puntaje_similitud)

In [14]:
# Ordenar los libros basados en su puntaje de similitud.
libros_similares_ordenados = sorted(puntaje_similitud, key=lambda x: x[1], reverse=True)
print(libros_similares_ordenados)

[(3373, 1.0), (3627, 0.705460825147957), (3628, 0.6925763481003818), (3372, 0.6800989802680497), (3374, 0.6537665447468091), (4289, 0.6419304070775775), (3626, 0.3876332117292275), (2529, 0.3870631474616968), (3364, 0.35298228448404007), (401, 0.34890082547406415), (3376, 0.3454818494585021), (175, 0.33850490637702374), (128, 0.33005239446550494), (120, 0.32166828120149693), (119, 0.3009930945573875), (300, 0.29562573057753405), (2334, 0.29366124139214433), (4558, 0.2905638613778546), (3368, 0.2813289293986784), (3494, 0.2743927905226895), (4576, 0.27364915904074205), (5599, 0.26876536032804116), (6592, 0.2584536790221493), (3366, 0.25616005589913327), (4574, 0.2499747487068429), (2602, 0.24384900325021888), (118, 0.2385985939555521), (5165, 0.22953005879062793), (537, 0.2213327931318359), (5570, 0.2199551119373641), (542, 0.2154634566157687), (5987, 0.21307956519238785), (4747, 0.21267291951156617), (1831, 0.2112660954960237), (1832, 0.2112660954960237), (3733, 0.2084435029693069), (3

In [15]:
# Elejimos los primeros n libros más similares al libro ingresado por el usuario.
n_libros_similares = 10

top_n_libros = libros_similares_ordenados[1 : n_libros_similares + 1]

top_n_libros

[(3627, 0.705460825147957),
 (3628, 0.6925763481003818),
 (3372, 0.6800989802680497),
 (3374, 0.6537665447468091),
 (4289, 0.6419304070775775),
 (3626, 0.3876332117292275),
 (2529, 0.3870631474616968),
 (3364, 0.35298228448404007),
 (401, 0.34890082547406415),
 (3376, 0.3454818494585021)]

In [16]:
# Impresión de los libros recomendados basados en el indice de los libros.

i = 1

print(f"Libro seleccionado por el usuario: {libro_usuario}\n")


print(f"Libros similares sugeridos por el sistema:\n")
for libro in libros_similares_ordenados:
    indice_libro = libro[0]
    titulo_indice = data[data.index == indice_libro]["title"].values[0]
    if i < n_libros_similares + 1:
        print(f"Libro {i}: {titulo_indice}")
        i += 1

Libro seleccionado por el usuario: peter pan

Libros similares sugeridos por el sistema:

Libro 1: Reaper Man
Libro 2: Soul Music
Libro 3: The Last Hero
Libro 4: Moving Pictures
Libro 5: Lords and Ladies
Libro 6: Rincewind the Wizzard
Libro 7: Eric
Libro 8: The Truth
Libro 9: Strata
Libro 10: The Light Fantastic


Para **evaluar** si la lista de libros recomendados es similar al libro ingresado y mostrar un porcentaje de similitud en cada ítem, vamos a calcular y mostrar el puntaje de similitud de cada libro recomendado en relación con el libro ingresado. Entonces vamos a:

- Calcular la Similitus del Coseno entre el libro ingresado y los libros recomendados.
- Agregar a la lista el porcentaje de similitud de cada libro recomendado en relación con el libro ingresado.
- Mostrar los libros recomendados ordenados por similitud.

A continuación la **Nueva Implementación** del sistema de recomendación de libros:

Nota: En esta función se agrega la variable subtitulo a la descripción del libro para mejorar la similitud entre los libros y retorna una lista para manipular posteriormente en el despliegue.Se controlaron posibilidades de que no encuentre coincidencias y se agregaron mensajes de error.

In [23]:
import difflib
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd

# Carga
url = "https://raw.githubusercontent.com/sosarodrigox/grupo13_pp2/main/data/data.csv"
data = pd.read_csv(url)

# Preprocesamiento
data.fillna("", inplace=True)
libros = (
    data["title"]
    + " "
    + data["subtitle"]
    + " "
    + data["categories"]
    + " "
    + data["authors"]
    + " "
    + data["published_year"].astype(str)
)

# Vectorización TF-IDF
vector_tfidf = TfidfVectorizer()
vector_caracteristicas = vector_tfidf.fit_transform(libros)

# Cálculo similitud de coseno
similitud = cosine_similarity(vector_caracteristicas, vector_caracteristicas)


def recomendar_libros(nombre_libro):
    print(f"Libro seleccionado por el usuario: {nombre_libro}")
    lista_titulos_completa = data["title"].tolist()
    encontrar_cercanos = difflib.get_close_matches(nombre_libro, lista_titulos_completa)

    if encontrar_cercanos:
        cercanos = encontrar_cercanos[0]
        indice_de_libro = data[data.title == cercanos].index[0]
        puntaje_similitud = list(enumerate(similitud[indice_de_libro]))
        libros_similares_ordenados = sorted(
            puntaje_similitud, key=lambda x: x[1], reverse=True
        )

        # Lista de diccionarios con los libros recomendados y su similitud
        recomendaciones = []
        # Muestra los 10 primeros
        for i, (index, score) in enumerate(libros_similares_ordenados[1:11], 1):
            libro = {
                "Indice": i,
                "Titulo": data.iloc[index]["title"],
                "Autor": data.iloc[index]["authors"],
                "Año": data.iloc[index]["published_year"],
                "Categoria": data.iloc[index]["categories"],
                "Similitud": f"{score * 100:.2f}%",
            }
            recomendaciones.append(libro)

        # Convertir a un DataFrame
        df_recomendaciones = pd.DataFrame(recomendaciones)
        return df_recomendaciones

    else:
        print("No se encontró ninguna coincidencia para el libro ingresado.")
        return pd.DataFrame()  # DataFrame vacío si no hay coincidencias


# Prueba
libro_usuario = input("Ingrese el nombre de su libro favorito: ")
df_recomendaciones = recomendar_libros(libro_usuario)
display(df_recomendaciones)

/var/folders/kf/yg24jn117tsddk3h_24784nr0000gp/T/ipykernel_15518/227369190.py:11: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  data.fillna('', inplace=True)


Libro seleccionado por el usuario: dracula


Indice                             Titulo  \
0       1                            Dracula   
1       2              Bram Stoker's Dracula   
2       3  The Bram Stoker Bedside Companion   
3       4               In Search of Dracula   
4       5                            Dracula   
5       6                 A Dracula Handbook   
6       7         Happy Hour at Casa Dracula   
7       8      Best Ghost and Horror Stories   
8       9                         Count Zero   
9      10              The Lord of the Rings   

                                                           Autor     Año  \
0                                                    Bram Stoker  2003.0   
1                                                   Harold Bloom  2003.0   
2                                                    Bram Stoker  1973.0   
3                               Raymond T. McNally;Radu Florescu  1994.0   
4                                                    Bram Stoker  1997.0   
5                                               Elizabeth Miller  2005.0   
6                                                   Marta Acosta  2006.0   
7  Bram Stoker;Richard Dalby;Stefan R. Dziemianowicz;S. T. Joshi  1997.0   
8                                                 William Gibson  2006.0   
9                                                                 2002.0   

                               Categoria Similitud  
0                                Fiction    74.80%  
1                    Juvenile Nonfiction    56.28%  
2                                Fiction    53.00%  
3              Biography & Autobiography    40.59%  
4                                Fiction    37.95%  
5                              Education    32.84%  
6                                Fiction    23.41%  
7                                Fiction    21.47%  
8                                Fiction    19.47%  
9  Baggins, Frodo (Fictitious character)    19.40%

### CRISP-DM 6. Despliegue:

# CONCLUSIONES DEL PROYECTO

## Proyecto de Recomendación de Libros basado en IA
Nuestro proyecto de recomendación de libros puede considerarse un proyecto de inteligencia artificial debido a las técnicas y algoritmos que emplea para procesar, analizar y recomendar libros basados en similitudes textuales. A continuación, detallamos las técnicas de IA utilizadas y cómo se relacionan con nuestro proyecto.

### * Procesamiento de Lenguaje Natural (NLP) 
El Procesamiento de Lenguaje Natural es un campo de la inteligencia artificial que se ocupa de la interacción entre las computadoras y el lenguaje humano. En nuestro proyecto, utilizamos técnicas de NLP para analizar y vectorizar el texto de los títulos, categorías, autores y años de publicación de los libros.

#### Técnicas y Algoritmos Utilizados:
- **TF-IDF (Term Frequency-Inverse Document Frequency)**:
  - **¿Qué es?**: TF-IDF es una técnica de NLP que se utiliza para convertir el texto en una representación numérica (vectores) que refleja la importancia de una palabra en un documento y en el corpus total.
  - **¿Por qué es IA?**: TF-IDF ayuda a identificar y ponderar las palabras más relevantes en un conjunto de documentos, facilitando así el análisis y la comparación de textos. Es una técnica fundamental en el procesamiento de lenguaje natural.

### * Algoritmos de Similitud y Medición de Distancia
Los algoritmos de similitud y medición de distancia son esenciales en IA para comparar y encontrar similitudes entre diferentes elementos.

#### Técnicas y Algoritmos Utilizados:
- **Similitud del Coseno (Cosine Similarity)**:
- **¿Qué es?**: La similitud del coseno mide la similitud entre dos vectores calculando el coseno del ángulo entre ellos. Es una medida de similitud muy utilizada en NLP.
- **¿Por qué es IA?**: Este algoritmo permite comparar documentos o textos en un espacio vectorial, ayudando a identificar documentos similares. Es ampliamente utilizado en motores de recomendación y búsqueda.

### * Sistema de Recomendación
Los sistemas de recomendación son una aplicación directa de la IA, diseñados para predecir la preferencia de un usuario por un ítem basado en información disponible.

#### Técnicas y Algoritmos Utilizados:
- **Filtrado Basado en Contenido (Content-Based Filtering)**:
- **¿Qué es?**: El filtrado basado en contenido recomienda ítems (en este caso, libros) que son similares a otros ítems que el usuario ha mostrado interés.
- **¿Por qué es IA?**: Este método utiliza modelos y algoritmos de IA para analizar las características de los ítems y encontrar similitudes, proporcionando recomendaciones personalizadas.

Finalizamos observando que nuestro proyecto de recomendación de libros emplea varias técnicas y algoritmos que se consideran parte del campo de la inteligencia artificial:

- **TF-IDF para vectorización de texto** (una técnica de NLP).
- **Similitud del Coseno para medir similitud entre textos**.
- **Sistema de recomendación basado en contenido** para proporcionar recomendaciones de libros.

Estas técnicas permiten analizar y comparar grandes volúmenes de texto de manera eficiente, identificar patrones y similitudes, y proporcionar recomendaciones personalizadas, lo que hace que nuestro proyecto sea un claro ejemplo de una aplicación de inteligencia artificial.
